File: services/source_tests.go
package services

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/fmind/fkf/core"
	"github.com/fmind/fkf/sources"
)

// SourceTestOutcome is the closed result vocabulary for one declared source verification hook.
type SourceTestOutcome string

const (
	SourceTestPassed SourceTestOutcome = "passed"
	SourceTestFailed SourceTestOutcome = "failed"
)

// SourceTestRequest selects enabled hooks by default, every declared hook with All, or the exact
// named sources regardless of their enabled state.
type SourceTestRequest struct {
	Targets []string
	All     bool
}

// SourceTestResult is one source hook execution. Captured stdout and provider stderr are never
// exposed; the reviewed argv and safe process status are enough to reproduce a failure.
type SourceTestResult struct {
	Source  string            `json:"source"`
	Enabled bool              `json:"enabled"`
	Command string            `json:"command"`
	Outcome SourceTestOutcome `json:"outcome"`
	Elapsed string            `json:"elapsed"`
	Error   string            `json:"error,omitempty"`
}

// SourceTestReport summarizes one bounded verification run.
type SourceTestReport struct {
	Base     string             `json:"base"`
	Sources  []SourceTestResult `json:"sources"`
	Passed   int                `json:"passed"`
	Failed   int                `json:"failed"`
	Complete bool               `json:"complete"`
	Elapsed  string             `json:"elapsed"`
}

// FailureSummary renders only reviewed source names, argv, and safe process diagnostics.
func (report *SourceTestReport) FailureSummary() string {
	lines := make([]string, 0, report.Failed)
	for _, result := range report.Sources {
		if result.Outcome != SourceTestFailed {
			continue
		}
		lines = append(lines, fmt.Sprintf("%s: %s (command: %s)", result.Source, result.Error, result.Command))
	}
	return strings.Join(lines, "\n")
}

// TestSources runs the selected trusted hooks sequentially in stable order. Hooks are optional;
// a default or --all run with none is a successful empty verification, while naming a source
// without a hook is an actionable configuration error.
func TestSources(ctx context.Context, base *Base, request SourceTestRequest) (*SourceTestReport, error) {
	started := base.Now()
	report := &SourceTestReport{Base: base.Root(), Sources: []SourceTestResult{}, Complete: true}
	targets, err := resolveSourceTestTargets(base, request)
	if err != nil {
		return report, err
	}
	if len(targets) == 0 {
		report.Elapsed = base.Now().Sub(started).Round(time.Millisecond).String()
		return report, nil
	}
	if err := base.RequireTrust(ctx); err != nil {
		return report, err
	}
	for _, source := range targets {
		if err := checkContext(ctx); err != nil {
			return report, err
		}
		resultStarted := base.Now()
		command := sources.BuildTestCommand(source, base.Env, base.Config.Sync.Timeout)
		result := SourceTestResult{
			Source: source.Name, Enabled: source.Enabled, Command: command.Display(), Outcome: SourceTestPassed,
		}
		_, runErr := base.Runner.Run(ctx, command)
		result.Elapsed = base.Now().Sub(resultStarted).Round(time.Millisecond).String()
		if runErr != nil {
			if ctxErr := ctx.Err(); ctxErr != nil && errors.Is(runErr, ctxErr) {
				return report, ctxErr
			}
			result.Outcome, result.Error = SourceTestFailed, runErr.Error()
			report.Failed++
			report.Complete = false
		} else {
			report.Passed++
		}
		report.Sources = append(report.Sources, result)
	}
	report.Elapsed = base.Now().Sub(started).Round(time.Millisecond).String()
	return report, nil
}

func resolveSourceTestTargets(base *Base, request SourceTestRequest) ([]*core.Source, error) {
	if request.All && len(request.Targets) > 0 {
		return nil, fmt.Errorf("%w: --all cannot be combined with source names", core.ErrConfig)
	}
	if len(request.Targets) == 0 {
		targets := make([]*core.Source, 0, len(base.Config.Sources))
		for _, name := range base.Config.SourceNames() {
			source := base.Config.Sources[name]
			if len(source.Test) > 0 && (request.All || source.Enabled) {
				targets = append(targets, source)
			}
		}
		return targets, nil
	}
	targets := make([]*core.Source, 0, len(request.Targets))
	seen := make(map[string]struct{}, len(request.Targets))
	for _, name := range request.Targets {
		if _, duplicate := seen[name]; duplicate {
			return nil, fmt.Errorf("%w: duplicate source %q; name each test target once", core.ErrConfig, name)
		}
		seen[name] = struct{}{}
		source, err := base.Source(name)
		if err != nil {
			return nil, err
		}
		if len(source.Test) == 0 {
			return nil, fmt.Errorf("%w: source %s declares no test hook in %s", core.ErrConfig, name, base.Config.Path)
		}
		targets = append(targets, source)
	}
	return targets, nil
}

File: src/fkf/source_tests.py
"""Sequential execution of optional, trust-gated source verification hooks."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum

from fkf.base import Base
from fkf.config import ConfigError, Source
from fkf.process import Cancellation, CommandCanceledError, display_argv
from fkf.source_runtime import build_test_command
from fkf.timeutil import DurationNS, format_duration
from fkf.trust import require_trust


class SourceTestOutcome(StrEnum):
    PASSED = "passed"
    FAILED = "failed"


@dataclass(frozen=True, slots=True)
class SourceTestRequest:
    targets: tuple[str, ...] = ()
    all: bool = False


@dataclass(frozen=True, slots=True)
class SourceTestResult:
    source: str
    enabled: bool
    command: str
    outcome: SourceTestOutcome
    elapsed: str
    error: str = ""


@dataclass(frozen=True, slots=True)
class SourceTestReport:
    base: str
    sources: tuple[SourceTestResult, ...]
    passed: int
    failed: int
    complete: bool
    elapsed: str

    def failure_summary(self) -> str:
        """Render only reviewed source names, argv, and safe runner diagnostics."""
        return "\n".join(
            f"{result.source}: {result.error} (command: {result.command})"
            for result in self.sources
            if result.outcome is SourceTestOutcome.FAILED
        )


def _elapsed(start: datetime, end: datetime) -> str:
    nanoseconds = round((end - start).total_seconds() * 1_000) * 1_000_000
    return format_duration(DurationNS(nanoseconds))


def _targets(base: Base, request: SourceTestRequest) -> tuple[Source, ...]:
    if request.all and request.targets:
        raise ConfigError("--all cannot be combined with source names")
    if not request.targets:
        return tuple(
            source
            for source in (base.config.sources[name] for name in base.config.source_names())
            if source.test and (request.all or source.enabled)
        )
    seen: set[str] = set()
    targets: list[Source] = []
    for name in request.targets:
        if name in seen:
            raise ConfigError(f"duplicate source {name!r}; name each test target once")
        seen.add(name)
        source = base.source(name)
        if not source.test:
            raise ConfigError(f"source {name} declares no test hook in {base.config.path}")
        targets.append(source)
    return tuple(targets)


def run_source_tests(
    base: Base,
    request: SourceTestRequest | None = None,
    *,
    cancel: Cancellation | None = None,
) -> SourceTestReport:
    """Run selected hooks in stable order and continue after safe hook failures."""
    if cancel is not None and cancel.is_set():
        raise CommandCanceledError("command canceled")
    started = base.now()
    request = SourceTestRequest() if request is None else request
    targets = _targets(base, request)
    if not targets:
        return SourceTestReport(str(base.root), (), 0, 0, True, _elapsed(started, base.now()))
    require_trust(base.config, cancel=cancel)
    results: list[SourceTestResult] = []
    passed = 0
    failed = 0
    for source in targets:
        if cancel is not None and cancel.is_set():
            raise CommandCanceledError("command canceled")
        result_started = base.now()
        command = build_test_command(source, base.environment, base.config.sync.timeout)
        try:
            base.runner.run(command, cancel=cancel)
        except CommandCanceledError:
            raise
        except Exception as error:
            failed += 1
            results.append(
                SourceTestResult(
                    source.name,
                    source.enabled,
                    display_argv(command.argv),
                    SourceTestOutcome.FAILED,
                    _elapsed(result_started, base.now()),
                    str(error),
                )
            )
        else:
            passed += 1
            results.append(
                SourceTestResult(
                    source.name,
                    source.enabled,
                    display_argv(command.argv),
                    SourceTestOutcome.PASSED,
                    _elapsed(result_started, base.now()),
                )
            )
    return SourceTestReport(str(base.root), tuple(results), passed, failed, failed == 0, _elapsed(started, base.now()))


__all__ = [
    "SourceTestOutcome",
    "SourceTestReport",
    "SourceTestRequest",
    "SourceTestResult",
    "run_source_tests",
]
